### PR TITLE
feat(torch): torch margins and vine distribution

### PR DESCRIPTION
**Pull request by tvatter**
Stacked on #280. PR 8 of the vine-distribution stack (§5.7 / §9 of the design).

## What this changes

`Vinedist` had no torch lane, so a `TorchVinecop` could not be combined with margins into a full distribution — the one configuration the torch backend exists for. This adds the two `nn.Module` halves.

- **`TorchMargin`** is a `MarginBase[Tensor]` that is *also* an `nn.Module`. `torch.distributions.Distribution` has no `.to(device)` and, held as a plain attribute, contributes **nothing** to `state_dict`. So the parameters are registered and the distribution is *rebuilt per call* from them — the same shape `TorchBicop` already uses for its interpolation grid. `.to("cuda")`, `state_dict` round-trips, and `backward()` into a margin's parameters all work as a consequence rather than as special cases.
- **`TorchMargin.from_distribution(factory, parameters=...)`** is the general entry point: any callable returning a `torch.distributions.Distribution`, plus the tensors it closes over. `icdf` falls back to bisection over `support` for the families that implement `cdf` but not `icdf` (`Gamma`, `Chi2`), which makes them usable margins; `Beta` and `StudentT` implement neither, and that is documented as a limitation rather than papered over.
- **`TorchVinedist`** is `Vinedist[Tensor]` plus `nn.Module`. It initializes the module half *before* handing over to the `_bind_dist` seam that PR #278 added for exactly this purpose, holds the margins in a `ModuleList`, and aliases `logpdf` as `log_prob` so torch users find the operation under the name they expect.

## Design decisions, and why

**Every margin must be an `nn.Module`, enforced with a `TypeError`.** The alternative — accept a SciPy margin and let it work — is the silent-failure case: SciPy raises on gradient-carrying tensors and, without `requires_grad`, returns a plain `numpy.ndarray`, so the graph detaches with no diagnostic and the loss simply stops learning the margins. Failing at construction names the offending margin instead.

**`from_data` refuses rather than inheriting.** `Vinedist.from_data` fits `Kde1dMargin` margins and a compiled `Vinecop` — neither of which this class can hold, and there is no torch marginal estimator to substitute. Inheriting it would return an object whose type contradicts its contents. The `NotImplementedError` names both real routes: fit on the NumPy lane, or assemble `TorchVinecop.from_data(u)` with `TorchMargin`s directly.

**`logpdf` stays canonical; `log_prob` is an alias on the torch subclass only.** The rest of the library spells it `logpdf`, and a second spelling in `core` would be the kind of vocabulary drift AGENTS.md warns about. On the torch subclass the alias is idiomatic and costs one line.

**Margins keep their own device placement.** `TorchVinedist.to(device)` moves the copula and every margin because they are all registered modules — no bespoke device plumbing, and no `device` attribute on the protocol (which `BicopLike` deliberately omits for the same reason: an `nn.Module`-backed object has no *intrinsic* placement).

## Verification

All runs CPU-pinned (`CUDA_VISIBLE_DEVICES=""`, asserted `torch.cuda.is_available() is False`) — this machine's GPU drives the maintainer's desktop session.

```
make check    # ruff + ruff-format + bandit + codespell + ty all pass
make docs     # after rm -rf docs/_generate docs/_build: build succeeded
make test     # 738 passed, 23 xfailed, 97% coverage
```

This PR adds **58** tests (`tests/test_torch_margin.py`, `tests/test_torch_vinedist.py`) and lands `src/pyvinecopulib/torch/vinedist.py` at **100%** coverage.

What the tests actually pin, rather than structure:

- **Parity with the NumPy lane** — `TorchVinedist.logpdf` / `pdf` / `cdf` against `Vinedist` holding the equivalent copula and margins, so any difference is the torch port's.
- **Autograd reaches the margin parameters** — `backward()` from `-logpdf.mean()` produces finite, non-zero `.grad` on the tensors a `TorchMargin` was built from, which is the property that motivates the whole module-rebuild design.
- **`state_dict` actually carries the margins** — a round-trip through `state_dict` / `load_state_dict` reproduces the same `logpdf`, which is precisely what a plain-attribute `Distribution` would fail.
- **`.to(device)` moves everything** — asserted on the copula and every margin, on CPU placement.
- **The two refusals are refusals** — non-`nn.Module` margins raise `TypeError`; `from_data` raises `NotImplementedError`. Both messages are asserted to name the alternative, not just to be non-empty.
- **`icdf` bisection** — `Gamma` and `Chi2` round-trip `cdf` ∘ `icdf` to grid tolerance, and `Beta` / `StudentT` raise rather than returning something plausible.
- **Pickling** — `TorchMargin` and `TorchVinedist` round-trip, matching the guarantee the rest of the library holds.

## Not in scope

- **Discrete margins on the torch lane** (PR 8b in the design, §5.6). `TorchBicop`'s difference quotients (`pdf_c_d` / `pdf_d_d`, the 2- and 4-corner h-function differences, the `5e-5` midpoint fallback) are the prerequisite and are deliberately deferred with the hooks pre-wired: `TorchVinedist` inherits the `var_type` / `cdf_left` plumbing from `Vinedist`, so nothing here has to change when they land.
- **A torch 1-d KDE margin.** `_interp` already supplies a differentiable grid `cdf` and a closed-form `icdf`; the 1-d local-likelihood *fitter* is genuinely new work.
- **End-to-end (non-IFM) training.** Nothing blocks it, but `Gamma.cdf` has no gradient w.r.t. its shape parameter, so joint copula-and-margin optimization is family-dependent in a way worth measuring before advertising.

### CODE REVIEW

**Comment by tvatter** [comment]
Superseded by #294. This chain was one pull request per commit; it is replaced by four feature-sized pull requests (#292–#295) rebased onto current `main`. The code is byte-identical — verified by diffing the consolidated tip against this chain's tip, which differs only in `CHANGELOG.md` (renumbered to the new pull requests), `AGENTS.md` (one added convention bullet) and the five files `main` itself changed in #267/#268.
